#326 | Display NFT Transfers

ezra-guzman opened the pull request:
# Fixes #326 

## Description
This PR adds NFT transfers into the transfers tab

## Test scenarios
- this PR needs to be tested locally because we need to spoof the user account
- run `git fetch --all && git checkout 326-display-nft-xfers`
- change `.env` to `NETWORK="mainnet"`
- change the following lines in files:
    - `history.ts:264` - `const userAddressLower = '0x13B745FC35b0BAC9bab9fD20B7C9f46668232607'.toLowerCase();`
    - `nfts.ts:135` - `const owner = '0x13B745FC35b0BAC9bab9fD20B7C9f46668232607';`
    - `EVMChainSettings.ts:381` - `const address = '0x13B745FC35b0BAC9bab9fD20B7C9f46668232607';`
    - `EVMChainSettings.ts:463` - `const url = 'v1/account/0x13B745FC35b0BAC9bab9fD20B7C9f46668232607/transfers';`
- run `yarn dev`
- go to http://localhost:8081
- sign in
- go to the transactions tab
- navigate transactions and see NFTs displayed (easiest way to see this is to do 100 rows per page and go forward a few pages)
- click an NFT
    - you should be sent to NFT detail page


Also, see this video demonstrating the hover effect working at different positions on the screen:
https://github.com/telosnetwork/telos-wallet/assets/11837820/d53370dc-7e67-4181-9194-d87256976544



## Checklist:
<!---
You can remove the items that are not relevant for your project.
-->
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have cleaned up the code in the areas my change touches
-   [x] My changes generate no new warnings
-   [ ] Any dependent changes have been merged and published in downstream modules
-   [x] I have checked my code and corrected any misspellings
-   [x] I have removed any unnecessary console messages
-   [x] I have included all english text to the translation file and/or created a new issue with the required translations for the currently supported languages
-   [x] I have tested for mobile functionality and responsiveness
-   [x] I have added appropriate test coverage

ezra-guzman commented on src/components/ExternalLink.vue:
```diff
@@ -42,11 +42,20 @@ const formattedText = computed(() => {
+        min-width: 0;
+        overflow: hidden;
+        white-space: nowrap;
+        text-overflow: ellipsis;
```
this makes links which are too long ellipsize, like very long NFT collection names

ezra-guzman commented on src/components/evm/TableControls.vue:
```diff
@@ -67,6 +67,8 @@ function changePageNumber(direction: 'next' | 'prev' | 'first' | 'last') {
         page,
     });
+
+    window.scrollTo(0, 0);
```
when the pagination info is changed, go to the top of the table

ezra-guzman commented on src/css/global/_colors.scss:
```diff
@@ -23,7 +23,7 @@ $saveBtn: #2e1f4f;
     $accent-3: #DDD1FF;
     $accent-4: #A5A6F6;
-    $page-header: #F7F5FF;
+    $accent-5: #F7F5FF;
```
update colors to match design system

ezra-guzman commented on src/layouts/EVMLayout.vue:
```diff
@@ -1,5 +1,25 @@
+
+const resizeListener = debounce(() => {
+    platformStore.setEvmMenuIsCollapsed(window.innerWidth < 1024);
+}, 100);
```
when window size changes, inform the store that the nav is either collapsed or open (large desktop vs smaller screens)

donnyquixotic reviewed (approved):
Some nft images do not display, is this expected?
![image](https://github.com/telosnetwork/telos-wallet/assets/6848910/58691d2d-9b93-4451-9c5f-89de78b919a3)

ezra-guzman commented:
Fixed a few things:
- all transactions with ERC20 transfers were calling the marketdata endpoint; price data is now cached
- a bug was causing all erc721 and erc1155 transfers to be re-fetched on pagination; now fixed
- the entire NFT transfers list was being compared to every transaction to determine which NFT transfers were a part of the transaction; NFT transfers are now stored in a Map keyed by transaction hash when fetched, so for each transaction, the comparison has been optimized to be a Map lookup

Viterbo commented on src/antelope/stores/history.ts:
```diff
@@ -179,42 +296,93 @@ export const useHistoryStore = defineStore(store_name, {
+
+                        nftDetailList.forEach((nftDetails, index) => {
+                            if (nftDetails) {
+                                let nftMediaType: 'image' | 'video' | 'audio' | 'unknown' = 'unknown';
+
+                                if (nftDetails.audioSrc) {
+                                    nftMediaType = 'audio';
+                                } else if (nftDetails.videoSrc) {
+                                    nftMediaType = 'video';
+                                } else if (nftDetails.imageSrcFull) {
+                                    nftMediaType = 'image';
+                                }
```
You can simply take the media type from the nftDirectly this way:
`const nftMediaType = nftDetails.item.type;`
The only problem would be that the 'unknown' value is not part of the current list of values. You can rename the NFTSourceTypes.NONE to NFTSourceTypes.UNKNOWN (which I think is the best) or you can adapt your code to use the 'none' value instead of 'unknown'

Viterbo commented on src/antelope/stores/history.ts:
```diff
@@ -316,4 +503,7 @@ const historyInitialState: HistoryState = {
         current: [],
     },
     __evm_transactions_pagination_data: {},
+    __evm_nft_transfers: {
+        current: new Map(),
+    },
```
just to be consistent
```suggestion
    __evm_nft_transfers: {
        [CURRENT_CONTEXT]: new Map(),
    },
```

Viterbo commented on src/pages/evm/nfts/NftViewer.vue:
```diff
@@ -113,19 +121,220 @@ function toggleVideoPlay(playOnly?: boolean) {
+    // the minimum padding between the hover element and the border of the window / the edge of the left navbar
+    const padding = 16;
+    // the width of the left nav
+    const leftNavWidth = platformStore.__evm_nav_is_collapsed ? 0 : 300;
```
Is it okay to directly access the variable this way or will we use a getter?

ezra-guzman replied:
yeah you can access state directly, thats fine. getters are just for more complex ways of accessing, like if we need a label or something

Viterbo reviewed (approved):
Really nice job, dude. 
I :heart: the viewer Component
